EIPS/eip-5388.md

```diff
@@ -0,0 +1,151 @@
+---
+eip: 5388
+title: Token-Gated HTTP Endpoints
+description: Composable RESTful and Solidity interface to implement token-gated HTTP endpoints using data tokens.
+author: Tim Daubenschütz (@TimDaub)
+discussions-to: https://ethereum-magicians.org/t/token-gated-https-endpoints/10205
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-08-01
+requires: 712
+---
+
+## Abstract
+
+This standard introduces composable RESTful HTTP and Solidity interfaces that enable paying to access HTTP endpoints using [EIP-20](./eip-20.md) tokens.
+
+## Motivation
+
+According to RFC 7231, HTTP status code `402 Payment Required` is reserved for future use. But the future is now as Ethereum enables users to verify their identity with a server using public key infrastructure.
+
+The Ethereum community has expressed the need for a composable method of token-gating HTTP endpoints. In this document, we outline a protocol implementable by HTTP servers to accept [EIP-20](./eip-20.md) payments prior to allowing access to an authorized endpoint.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+Upon the user's initial request, an HTTP server implementing token-gating via this EIP responds with a `402 Payment Required` status code:
+
+- Its body must include a valid JSON string that complies with the Solidity Contract ABI Specification for calling contracts.
+- It must include a custom header `X-EIP-5388-CONTRACT-ADDRESS` that suggests the contract's address a call must be directed to.
+- It must include a custom header `X-EIP-5388-ABI-ENCODED-INPUT` that suggests the call signature the user must invoke at `X-EIP-5388-CONTRACT-ADDRESS` to gain access to the endpoint.
```

Comment:
I feel the spec also lacks a part in solidity interface for what smart contract it will be to respond to a end point request

Reply:
not sure I understand what you mean, would you mind elaborating?

Reply:
I think what @xinbenlv is asking is probably which are these contracts that these endpoints might be dependent on, as suggested in his changes. Is that it?